### PR TITLE
post: ng perf series part 5

### PR DESCRIPTION
**Pull request by cexbrayat**
Final part of the series

### CODE REVIEW

File: _posts/2018-10-04-angular-performances-part-5.md [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+
+What you may not know is that using a pipe is also more performant.
+By default, a pipe is "pure".
+In computer science, we call "pure" a function that has no side effect,
+and which result only depends on its entries.
+A pure pipe is pretty much the same:
+the result of its `transform` method only depends on arguments.
+Knowing that, Angular applies a nice optimization:
+the `tranform` method is only called if the reference of the value it transforms changes
+or if one of the other arguments change
```
Comment (clacote) [inline _posts/2018-10-04-angular-performances-part-5.md]: change -> changes (one argument)

File: _posts/2018-10-04-angular-performances-part-5.md [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+This is a simple pattern to use,
+often referred to as the smart/dumb component pattern:
+a smart component deals with data loading, event handling, etc.,
+and simply passes the data to display as input to a second, dumb component.
+The only responsibility of the dumb component is to display the data,
+and to emit events to its parent smart component using outputs.
+This dumb component is the one with the large template,
+containing many expressions.
+But since its state only changes when its smart parent passes a new input,
+it can use OnPush and thus save a lot of expression evaluations.
```
Comment (clacote) [inline _posts/2018-10-04-angular-performances-part-5.md]: save -> saves (it, the sub-component)

File: _posts/2018-10-04-angular-performances-part-5.md [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+      transform(user: UserModel): string {
+        return `${user.title} ${user.firstName} ${user.lastName}`;
+      }
+
+    }
+
+To sum up:
+
+- a pure pipe is not called as often as a method in a component
+- but it doesn't run again if the input value is mutated, so use carefully
```
Comment (clacote) [inline _posts/2018-10-04-angular-performances-part-5.md]: Add a final `.`?

**Comment by cexbrayat** [comment]
This will be published tomorrow morning (last chance for review @jnizet )

File: _posts/2018-10-04-angular-performances-part-5.md [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+    }
+{% endraw %}
+
+This takes a little bit more work,
+but writing a pipe allows to reuse it in other components.
+
+What you may not know is that using a pipe is also more performant.
+By default, a pipe is "pure".
+In computer science, we call "pure" a function that has no side effect,
+and which result only depends on its entries.
```
Comment (jnizet) [inline _posts/2018-10-04-angular-performances-part-5.md]: I would have said "whose result", but maybe I'm wrong.

File: _posts/2018-10-04-angular-performances-part-5.md [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+and only once in a template if it is used with the same input value and arguments (as in my example).
+
+By default, a custom pipe is pure, so that's great!
+But sometimes it's not a right fit.
+
+In my example, if we mutate the user to set its `firstName` to a different value,
+the pipe never refreshes...
+It's pretty much the same issue that we had with the `OnPush` strategy:
+the reference of the value doesn't change,
+so pipe does not run again.
```
Comment (jnizet) [inline _posts/2018-10-04-angular-performances-part-5.md]: so **the** pipe

**Pull request merged**